File: pacman-reloaded/src/main/java/scenes/pacman_scene.rb
require 'java'
java_import Java::com.uqbar.vainilla.GameScene
java_import Java::extensions.shapes.Circle
require '../pacman/pacman_game'


class PacmanScene < GameScene
  attr_accessor :walkable_matrix, :block_size

  def onSetAsCurrent
    super
    image = Resource.get_image 'pacman-min.png'
    shape = Circle.new image.width / 2, 0, 0
    pacman = Pacman.new(shape, image, 0 * block_size, 1 * block_size)
    pacman.initialize_vector 1, 0, 200
    add_component pacman
  end

  def can_walk(x, y)
    walkable_matrix[y][x]
  end

  def can_occupy?(x, y, circle)
    can_walk to_b(x), to_b(y) and
        can_walk to_b(x + circle.diameter), to_b(y) and
        can_walk to_b(x), to_b(y + circle.diameter) and
        can_walk to_b(x + circle.diameter), to_b(y + circle.diameter)

  end

  def to_b(x)
    (x / block_size).to_i
  end

end

File: pacman-reloaded/src/main/java/scenes/PacmanScene.java
package scenes;


import extensions.shapes.Circle;
import extensions.shapes.Shape;

import java.util.ArrayList;
import java.util.List;

public class PacmanScene extends AbstractPacmanScene {
    private List<List<Boolean>> list = new ArrayList<List<Boolean>>();

    public boolean canOccupy(double  x, double y, Circle circle) {
        return can_walk(to_b(x), to_b(y)) && can_walk(to_b(x + circle.getDiameter()), to_b(y)) &&
                can_walk(to_b(x), to_b(y + circle.getDiameter())) &&
                can_walk(to_b(x + circle.getDiameter()), to_b(y + circle.getDiameter()));
    }

    public int to_b(double coord) {
        return (int) (coord / block_size());
    }

    public int block_size() {
        return getGame().getConfig().getInt("blockSize");
    }

    public boolean can_walk(int x, int y) {

        return getList().get(y).get(x);
    }


    public List<List<Boolean>> getList() {
        return list;
    }

    public void setList(List<List<Boolean>> list) {
        this.list = list;
    }
}
